ui/components/resources-grant.js

```diff
@@ -0,0 +1,126 @@
+import { useState } from 'react'
+import useSWR, { mutate } from 'swr'
+import { PlusIcon } from '@heroicons/react/outline'
+
+import InputDropdown from '../components/input'
+import ErrorMessage from '../components/error-message'
+
+export default function ({ id }) {
+  const { data: grants } = useSWR(`/v1/identities/${id}/grants`)
+  const { data: destinations } = useSWR('/v1/destinations')
+
+  const [infrastructure, setInfrastructure] = useState('')
+  const [error, setError] = useState('')
+  const [role, setRole] = useState('view')
+
+  const options = ['view', 'edit', 'admin', 'remove']
+
+  const handleInputChang = value => {
+    setInfrastructure(value)
+    setError('')
+  }
+
+  const handleKeyDownEvent = key => {
+    if (key === 'Enter' && infrastructure.length > 0) {
+      handleShareGrant()
+    }
+  }
+
+  const handleShareGrant = async () => {
+    if (destinations.find(d => d.name === infrastructure)) {
+      grant(id)
+    } else {
+      setError('TODO: the infrastructure does not exist')
+    }
+  }
+
+  const grant = (user, privilege = role, resource = infrastructure, exist = false, deleteGrantId) => {
```

Comment:
nit: `function grant (user, ...)` might be a bit cleaner – this is just personal preference though